C++ migration: add libdispatch-based version of FSTDispatchQueue

var-const opened the pull request.

var-const commented on Firestore/core/test/firebase/firestore/util/async_queue_test.cc:
```diff
@@ -0,0 +1,248 @@
+  std::string steps;
+
+  auto queue = Queue();
+  queue.Enqueue([&steps] { steps += '1'; });
```
My reading of the [docs](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man3/dispatch_async_f.3.html) is that accessing `steps` from several enqueued tasks is thread-safe:

> The dispatch_async() function may be used to implement deferred critical sections [...] queues support both asynchronous and synchronous operations, and those operations are ordered by definition.

But I'd appreciate if @wilhuff could confirm once he gets back from vacation.

var-const commented on Firestore/core/src/firebase/firestore/util/async_queue.h:
```diff
@@ -0,0 +1,204 @@
+ * AsyncQueue. These ids can then be used from tests to check for the
+ * presence of delayed operations or to run them early.
+ */
+enum class TimerId {
```
The name is the same on other platforms, but I'd suggest `Tag` or `TimerTag` -- really, the operations are tagged by this.

zxu123 commented on Firestore/core/src/firebase/firestore/util/async_queue.h:
```diff
@@ -0,0 +1,204 @@
+  void RunSync(const Operation& operation);
+
+  /**
+   * For tests: determine if a delayed operation with a particular `TimerId`
```
Just my 1 cent, those helper methods purely for tests can be defined `private` and make test friend class rather than expose them `public`. From maintenance point of view, change private methods is easier that changing a public one, the latter of which may involve larger scope of callers.

rsgowman replied:
Mental exercise: It might be neat to create a 'test' namespace nested in this one, with a 'DelayedOperationTester' class. You could make these methods private here, and mirror them as public in the Tester. It'd make it more obvious that they exist only for testing purposes and shouldn't be otherwise relied on.

Optional.

var-const replied:
I completely agree with the general idea, but I don't think I will have enough context to figure out what the best solution is until I get to changing Objective-C code to using this version of queue.

Another thing to note: the vector of `DelayedOperation`s exists mostly for the purpose of tests. For cancellation, just transferring ownership of `DelayedOperation` libdispatch should be enough.

zxu123 commented on Firestore/core/src/firebase/firestore/util/async_queue.h:
```diff
@@ -0,0 +1,204 @@
+  dispatch_queue_t dispatch_queue_{};
+  using OperationPtr = std::shared_ptr<detail::DelayedOperationImpl>;
+  std::vector<OperationPtr> operations_;
+  bool is_operation_in_progress_{};
```
might just explicit set the initial value `true` or `false` than default initialize {}, which do have defined meaning of set to `false` but harder for human to read.

zxu123 commented on Firestore/core/src/firebase/firestore/util/async_queue.cc:
```diff
@@ -0,0 +1,314 @@
+// interface: accepts an arbitrary invocable object in place of an Objective-C
+// block.
+template <typename Work>
+void DispatchAsync(dispatch_queue_t queue, Work&& work) {
```
make queue `const` as well?

zxu123 commented on Firestore/core/src/firebase/firestore/util/async_queue.cc:
```diff
@@ -0,0 +1,314 @@
+  void HandleDelayElapsed() {
+    queue_->VerifyIsCurrentQueue();
+    if (!done_) {
+      done_ = true;
```
Could there be race case that leads operation being run twice? I guess not if we can guarantee that a DelayedOperationImpl cannot be put into two different dispatch-queue ever.

zxu123 commented on Firestore/core/src/firebase/firestore/util/async_queue.cc:
```diff
@@ -0,0 +1,314 @@
+                                            AsyncQueue::Milliseconds>;
+
+  AsyncQueue* queue_{};
+  TimerId timer_id_{};
```
Might just initialize the enum TimerId directly.

zxu123 commented on Firestore/core/src/firebase/firestore/util/async_queue.cc:
```diff
@@ -0,0 +1,314 @@
+  using TimePoint = std::chrono::time_point<std::chrono::system_clock,
+                                            AsyncQueue::Milliseconds>;
+
+  AsyncQueue* queue_{};
```
Same here, `queue_ = nullptr` is easier to read than `queue_{}`

zxu123 commented on Firestore/core/src/firebase/firestore/util/async_queue.cc:
```diff
@@ -0,0 +1,314 @@
+  return DelayedOperation{operations_.back()};
+}
+
+void AsyncQueue::RunSync(const Operation& operation) {
```
Is this implementation identical to `AsyncQueue::Enqueue`? If logically the same, might let one call the other or just keep one.

zxu123 replied:
That's right. I was mentally auto-corrected sync to async for the `AsyncQueue` class and now realized this meant to run it syncly.

rsgowman commented on Firestore/core/test/firebase/firestore/util/async_queue_test.cc:
```diff
@@ -0,0 +1,246 @@
+
+    DelayedOperation delayed_operation = queue.EnqueueAfterDelay(
+        AsyncQueue::Milliseconds(1), kTimerId1, [&steps] { steps += '2'; });
+
```
(as above)

rsgowman commented on Firestore/core/test/firebase/firestore/util/async_queue_test.cc:
```diff
@@ -0,0 +1,246 @@
+  auto queue = Queue();
+  DelayedOperation delayed_operation = queue.EnqueueAfterDelay(
+      AsyncQueue::Milliseconds(1), kTimerId1, [&] { signal_finished(); });
+  EXPECT_TRUE(queue.ContainsDelayedOperation(kTimerId1));
```
(as above)

rsgowman commented on Firestore/core/test/firebase/firestore/util/async_queue_test.cc:
```diff
@@ -0,0 +1,246 @@
+
+  auto queue = Queue();
+  queue.Enqueue([&steps] { steps += '1'; });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(20000), kTimerId1,
```
technically, as above. But if we're context switching for 10s+, we've got bigger problems.

zxu123 commented on Firestore/core/src/firebase/firestore/util/async_queue.h:
```diff
@@ -0,0 +1,204 @@
+
+namespace detail {
+class DelayedOperationImpl;
+}
```
Actually, it seems that we are using `impl` for such case in addition to `internal`, see `git grep "namespace impl"` and `git grep "namespace internal"`.

rsgowman replied:
fwiw searching cs for "namespace detail|internal|impl" shows ratios of approx 2:30:1 respectively.

var-const commented on Firestore/core/src/firebase/firestore/util/async_queue.h:
```diff
@@ -0,0 +1,204 @@
+#include <vector>
+
+#include "absl/strings/string_view.h"
+#include "dispatch/dispatch.h"
```
CMake build runs okay on my Mac. (by the way, it's great you're running the build on Linux, it's not the first time you notice a problem that isn't apparent on a Mac)

var-const commented on Firestore/core/test/firebase/firestore/util/async_queue_test.cc:
```diff
@@ -0,0 +1,246 @@
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+  EXPECT_EQ(steps, "1234");
+}
```
I don't think that's a problem, but I may misunderstand what you mean. My thinking is:
* libdispatch guarantees the relative order of the dispatched work. That is, work scheduled with 1ms delay is guaranteed (AFAICT) to be run before the work scheduled with 5ms delay. The queue the tests are using is a serial queue, so the tasks are run one at a time.
* if libdispatch takes too long, I don't see how a fake clock can help.
* I'm also not sure how exactly libdispatch checks time (to be more precise, whether it's well-documented).

So I don't think there should be a problem - the test basically checks the correct ordering, and isn't really timing-dependent (well, it can time out, but I think it's an unavoidable problem). Please let me know if I'm missing something.

rsgowman replied:
I meant the following:

```
steps = '1';
queue.Enqueue(delay=4, fn=() { steps += '4';}
queue.Enqueue(delay=3, fn=() { steps += '3';}
queue.Enqueue(delay=2, fn=() { steps += '2';}
```

Expected output would be "1234". (Good.) But now consider:

```
steps = '1';
queue.Enqueue(delay=4, fn=() { steps += '4';}
MasiveContextSwitchCausingDelayOfAtLeast5();
queue.Enqueue(delay=3, fn=() { steps += '3';}
queue.Enqueue(delay=2, fn=() { steps += '2';}
```

In this case, could the output be "1423"?

If libdispatch doesn't actually start processing things in the queue until WaitForTestToFinish(), then no action required. But if it could start processing as soon as something is added to the queue, it seems like you could end up in trouble.


> if libdispatch takes too long, I don't see how a fake clock can help.

Not due to processing, but due to sleeps. For instance, this test will take at least 5ms to run. (Which is pretty quick. But imagine if the test is actually flakey and a temporary measure to fix that causes us to increase that to 5s.) Sitting around and waiting for those timeouts/delays to expire isn't awful for integration tests, but is unfortunate for unit tests.

```
steps = '1';
queue.Enqueue(delay=4, fn=() { steps += '4';}
queue.Enqueue(delay=3, fn=() { steps += '3';}
queue.Enqueue(delay=2, fn=() { steps += '2';}
fakeClock.Advance(5);
WaitForQueueToFinish(); // still necessary, since we need to allow the queue to notice the time and run async tasks. But should complete more-or-less immediately.
```

Of course, that all assumes you *can* inject a fake clock into libdispatch. If it has a hard dependency on the system clock then there's probably not much that can reasonable be done.

var-const replied:
> In this case, could the output be "1423"?

It's a serial queue. Unless I have a pretty big misunderstanding of how libdispatch works, it guarantees relative order of dispatched operations. The next dispatch call on a serial queue is only ever made when the previous one has finished. So I don't see how this could be a problem.

> For instance, this test will take at least 5ms to run.

All the actual code that uses this functionality has delay of several seconds. I chose milliseconds granularity for the sake of tests pretty much arbitrarily. It could be nanoseconds, what do you think?

> Of course, that all assumes you can inject a fake clock into libdispatch. If it has a hard dependency on the system clock then there's probably not much that can reasonable be done.

I'm pretty skeptical whether it's possible, and whether it's something worth relying upon.

rsgowman replied:
> So I don't see how this could be a problem.

Lets sync in person. (I've got a repro for my concern locally. Whether it's a real concern or not...?)

> I chose milliseconds granularity for the sake of tests pretty much arbitrarily. It could be nanoseconds, what do you think?

Oh, as long as there's no race, I don't care. 5ms won't be our biggest bottleneck. :)

var-const commented on Firestore/core/src/firebase/firestore/util/async_queue.h:
```diff
@@ -0,0 +1,204 @@
+  friend class AsyncQueue;
+  explicit DelayedOperation(
+      const std::shared_ptr<detail::DelayedOperationImpl>& instance)
+      : handle_{instance} {
```
Renamed `instance` to `operation`, but I'm inclined to keep `handle`. I think it's important for the name to reflect that it doesn't own the operation and only has a handle to it. Feel free to suggest alternatives. Maybe `observer_`?

rsgowman commented on Firestore/core/src/firebase/firestore/util/CMakeLists.txt:
```diff
@@ -105,6 +105,8 @@ configure_file(
 cc_library(
   firebase_firestore_util
   SOURCES
+    async_queue.cc
```
> [... ] The overall pattern in this folder's CMakeLists.txt is to build the resulting firebase_firestore_util library from several libraries, some of which are platform-dependent. I cannot easily do the same for the library containing AsyncQueue, because [...]

You could put them all into one library; but you'd have to add the non-apple libdispatch now-ish. Assuming you did that, you could:
1) Add the external libdisaptch as a conditional dependency, i.e.
```
if(not apple)
  list(append UTIL_DEPENDS libdispatch_dispatch_or_whatever_we_call_it)
endif()
```
2) add async_queue_libdispatch.* to SOURCES of firebase_firestore_util

Alternatively, putting this into a separate library or directory would also make sense, but doesn't mirror the obj-c code. (Which is ok, but should trigger further discussion with the larger team.) If you take this approach, rather than defining the library conditionally, I'd recommend adding the dependency conditionally. (See appendix below.)

As a temporary measure, defining the library only on apple (as you've done) is fine by me for now, though I'd recommend a TODO be added.


> Also, what async_queue library does is: [...]

One could imagine more than just the sources differing (i.e. dependencies, cflags, etc). In the extreme, this would cause the second approach to mirror the first, but with more variables. (But I would imagine that this would not be a common case.)

Looking a f/f/{auth,util}/CMakeLists.txt, we seem to be defining "private" libraries that are platform specific, and then conditionally adding them to the dependencies of "public" libraries. In other words, it would look more like this:

```
cc_library(apple_specific_thing, ...)
cc_library(nonapple_thing, ...)

if (apple)
  list(append deps apple_specific_thing)
else
  list(append deps nonapple_thing)

cc_library(
  publicish_lib
  SOURCES
    additional_platform_independent_sources.cc
  DEPENDS
    ${deps}
    additional_platform_independent_deps
)
```

Of note, the apple library would end up defined, even on non-apple platforms, though it won't actually be compiled. (This ensures the sources are present, I think, but doesn't require them to be compile-able.)

That may all be less relevant here, since I believe the source will be same on all platforms?
 
 
Appendix: Yeah, this comment is now long enough to warrant an appendix. :(

If you want to define this as a separate library (rather than being part of f_f_util), I'd recommend:

1) Create libdispatch_dispatch library (as an external dependency. Check out cmake/external/*.cmake and cmake/Find*.cmake)
2) Add the following to f/f/util/CMakeLists.txt
```
if (not apple)
  list(append async_queue_deps libdispatch_dispatch)
endif()
```
3) alter f/f/util/CMakeLists.txt and unconditionally define firebase_firestore_util_async_queue.
4) Add a ${async_queue_deps} to the DEPENDS section

var-const replied:
> you'd have to add the non-apple libdispatch now-ish

I don't intend to do that. I think the idea is to only use libdispatch-based implementation on Apple platforms, at least for now.

So far I have no platform-independent source files for this (the C++-based version is completely separate). Maybe it will be possible to refactor out a common part that delegates to a platform-specific "impl", but I'm not sure at this point.

rsgowman replied:
Oh, I misunderstood then; this should get easier as a result. I had assumed that this code would be used everywhere (and that we'd pull in a non-apple implementation of libdispatch such as https://github.com/nickhutchinson/libdispatch)

To be clear, this code will only ever be used on apple, right?

If so, then I think you can just
a) define the library (unconditionally)
b) conditionally add to util_depends:
```
if (apple)
  list(append util_depends f_f_u_async_queue)
endif()
```

var-const commented on Firestore/core/src/firebase/firestore/util/async_queue.h:
```diff
@@ -0,0 +1,204 @@
+   * @param operation The operation to run.
+   * @param delay The delay after which to run the operation.
+   * @param timer_id A `TimerId` that can be used from tests to check for the
+   *     presence of this operation or to schedule it to run early.
```
Rephrased the comment, PTAL.

> nit: I'm not sure that's technically correct since the implementation of EnqueueAfterDelay will assert if there's two operations with the same timer_id

Right.

> Curiously, only obj-c has this parameter. Android and typescript do not

They both have it, e.g. [TypeScript](https://github.com/firebase/firebase-js-sdk/blob/0fa319e5e019dd0d40ab441d2ff9f8f6d4724e43/packages/firestore/src/util/async_queue.ts#L237).

rsgowman replied:
> Rephrased the comment, PTAL.

I don't see it? (Last commit I see is from yesterday.)

Oh, my mistake. My client was out of date. By a lot. :)

var-const replied:
Sorry, it's actually my mistake. Pushed the commit now.

rsgowman replied:
I see the comment now. lgtm.

var-const commented on Firestore/core/src/firebase/firestore/util/async_queue_libdispatch.h:
```diff
@@ -0,0 +1,214 @@
+#include <vector>
+
+#include "absl/strings/string_view.h"
+#include "dispatch/dispatch.h"
```
Hmm, linker complains: "Firestore/core/src/firebase/firestore/util/async_queue_libdispatch.h:20:  <dispatch/dispatch.h> should be #include "dispatch/dispatch.h" or #import <dispatch/dispatch.h>  [build/include] [4]"

wilhuff commented on Firestore/core/src/firebase/firestore/util/async_queue_libdispatch.h:
```diff
@@ -17,7 +17,7 @@
 #define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_ASYNC_QUEUE_LIBDISPATCH_H_
 
-#include <dispatch/dispatch.h>
+#include "dispatch/dispatch.h"
```
This is likely because the linter doesn't know about this and it's too conservative. If you see cases like this, please just add them to `_C_SYSTEM_DIRECTORIES` in cpplint.py. Once we've completed our migration I'll relax the check

wilhuff commented:
Re race on destruction: could this be resolved by destroying the operations array (and anything else sensitive like this) in a dispatch_sync?

var-const commented:
> Re race on destruction: could this be resolved by destroying the operations array (and anything else sensitive like this) in a dispatch_sync?

I have a theory (currently trying it out): perhaps waiting on the `packaged_task::get_future()` actually does provide the synchronization I expect, but the problem is that the worker thread accesses the vector (`RemoveDelayedOperation`) after the `packaged_task` is complete and thus this access is not synchronized. I'll try to remove the operation before running it and see if it helps.

var-const commented:
> the problem is that the worker thread accesses the vector (RemoveDelayedOperation) after the packaged_task is complete and thus this access is not synchronized.

It looks like I was right on this. Progress update: I reworked `EnqueueAfterDelay` to verify it's running on the worker queue, and changed the tests to respect that. Right now I'm working on fixing`RunDelayedOperationsUntil` (I didn't take note that in Objective-C, the callback object would still be alive by the time libdispatch gets to invoke it after the "force" invocation).

var-const commented:
Note: I'm in the process of reworking this PR to use (more) shared pointers. (the pure C++ version doesn't depend on this PR in any way)

var-const commented:
Reworked the ownership model in the PR, PTAL.

var-const commented:
To summarize the offline discussion with @wilhuff:
* I will try to simplify the ownership model. The main idea is to give libdispatch exclusive ownership of DelayedOperations (right now both libdispatch and AsyncQueue have equal ownership of those). Hopefully no issue comes up with this;
* I'll hold off both libdispatch and C++11 versions of AsyncQueue to refactor them to a "shared class with Firestore-specific logic/platform-dependent implementation with lower-level stuff)" model.
